flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecWatermarkAssigner.scala

```diff
@@ -62,37 +63,36 @@ class StreamExecWatermarkAssigner(
 
   override def requireWatermark: Boolean = false
 
-  override def copy(traitSet: RelTraitSet, inputs: util.List[RelNode]): RelNode = {
-    new StreamExecWatermarkAssigner(
-      cluster,
-      traitSet,
-      inputs.get(0),
-      rowtimeFieldIndex,
-      watermarkDelay)
+  override def copy(
+      traitSet: RelTraitSet,
+      input: RelNode,
+      rowtime: Int,
+      watermark: RexNode): RelNode = {
+    new StreamExecWatermarkAssigner(cluster, traitSet, input, rowtime, watermark)
   }
 
+  /**
+    * Fully override this method to have a better display name of this RelNode.
+    */
   override def explainTerms(pw: RelWriter): RelWriter = {
-    val miniBatchInterval = traits.getTrait(MiniBatchIntervalTraitDef.INSTANCE).getMiniBatchInterval
-
-    val value = if (miniBatchInterval.mode == MiniBatchMode.None ||
-      miniBatchInterval.interval == 0) {
-      // 1. redundant watermark definition in DDL
-      // 2. existing window aggregate
-      // 3. operator requiring watermark, but minibatch is not enabled
-      "None"
-    } else if (miniBatchInterval.mode == MiniBatchMode.ProcTime) {
-      val tableConfig = cluster.getPlanner.getContext.unwrap(classOf[FlinkContext]).getTableConfig
-      val miniBatchLatency = getMillisecondFromConfigDuration(tableConfig,
-        ExecutionConfigOptions.TABLE_EXEC_MINIBATCH_ALLOW_LATENCY)
-      Preconditions.checkArgument(miniBatchLatency > 0,
-        "MiniBatch latency must be greater that 0 ms.", null)
-      s"Proctime, ${miniBatchLatency}ms"
-    } else if (miniBatchInterval.mode == MiniBatchMode.RowTime) {
-      s"Rowtime, ${miniBatchInterval.interval}ms"
-    } else {
-      throw new TableException(s"Unsupported mode: $miniBatchInterval")
-    }
-    super.explainTerms(pw).item("miniBatchInterval", value)
+    val inFieldNames = inputRel.getRowType.getFieldNames.toList
+    val rowtimeFieldName = inFieldNames(rowtimeFieldIndex)
+    val miniBatchInterval = traits
```

Comment:
It seems watermark assigner is influenced heavily by mini batch assigner, I'm not sure whether this is right way to creating physical operators. Normally all physical operators should be independent from each other.

Reply:
I think `WatermarkAssigner` is independent with `MiniBatchAssigner`. It is the `MiniBatchIntervalTrait` related with these 2 nodes. However, the `MiniBatchIntervalTrait` indeed should cooperate with watermark and without watermark. 

In the past, `WatermarkAssigner` and `MiniBatchAssigner` stayed in a single node, which makes the code complex and member fields nullable and hard to maintain.

Reply:
IIUC, `StreamExecMiniBatchAssigner` is only responsible for mini-batch assigner on **pro-time**, and `StreamExecWatermarkAssigner` is responsible for simple watermark assigner on   **pro-time**/**row-time**  and mini-batched watermark assigner both on **row-time** ?  

how about `StreamExecMiniBatchAssigner` is just responsible for mini-batch assigner (on **pro-time**/**row-time**) and `StreamExecWatermarkAssigner` is responsible for simple watermark assigner (on **pro-time**/**row-time**) ?